ducktape/cluster/node_container.py

```diff
@@ -103,25 +122,58 @@ def remove_nodes(self, nodes):
         for node in nodes:
             self.remove_node(node)
 
-    def remove_spec(self, cluster_spec):
+    def remove_spec(self, cluster_spec) -> RemoveSpecResult:
         """
         Remove nodes matching a ClusterSpec from this NodeContainer.
 
         :param cluster_spec:                    The cluster spec.  This will not be modified.
-        :returns:                               A list of the nodes that were removed.
-        :throws InsufficientResourcesError:     If there are not enough nodes in the NodeContainer.
-                                                Nothing will be removed unless enough are available.
+        :returns:                               An instance of `AllocResult`, which contains good nodes that passed
+                                                health checks, bad nodes that didn't, overall success/failure and
+                                                a detailed message (if any).
+        :raises:                                InsufficientResourcesError when there aren't enough total nodes
+                                                InsufficientHealthyNodesError when there aren't enough healthy nodes
         """
-        msg = self.attempt_remove_spec(cluster_spec)
-        if len(msg) > 0:
-            raise InsufficientResourcesError("Not enough nodes available to allocate. " + msg)
-        removed = []
+
+        err = self.attempt_remove_spec(cluster_spec)
+        if err:
+            # there weren't enough nodes to even attempt allocations, raise the exception
+            raise InsufficientResourcesError(err)
+
+        r = RemoveSpecResult()
+        msg = ""
+        # we have enough nodes for each OS, now try allocating while doing health checks if nodes support them
         for os, node_specs in iteritems(cluster_spec.nodes.os_to_nodes):
             num_nodes = len(node_specs)
+            good = []
             avail_nodes = self.os_to_nodes.get(os, [])
-            for i in range(0, num_nodes):
-                removed.append(avail_nodes.pop(0))
-        return removed
+            # loop over all available nodes
+            # for i in range(0, len(avail_nodes)):
+            while avail_nodes and (len(good) < num_nodes):
```

Comment:
Personal preference for obvious-ness - it's easier to read because it's VERY obvious about the priority of operations - without the parenthesis you have to actually think whether `and` or `<` has priority